File: encryption_utils/Go/decrypt.go
package main

import (
	"bufio"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/hex"
	"encoding/pem"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"path/filepath"
	"regexp"
	"strings"
)

var private, encryptedKey, encryptedfilepath string

func init() {
	flag.StringVar(&encryptedKey, "key", "", "encrypted symmetric key used for file decryption (hex-encoded string)")
	flag.StringVar(&encryptedfilepath, "file", "", "location of encrypted file")
	flag.StringVar(&private, "pk", "", "location of private key to use for decryption of symmetric key")
	flag.Parse()

	var missingArgs []string
	if encryptedKey == "" {
		missingArgs = append(missingArgs, "key")
	}
	if encryptedfilepath == "" {
		missingArgs = append(missingArgs, "file")
	}
	if private == "" {
		missingArgs = append(missingArgs, "pk")
	}
	if len(missingArgs) > 0 {
		fmt.Printf("Missing argument(s): %s\n", strings.Join(missingArgs, ", "))
		os.Exit(1)
	}

	r, _ := regexp.Compile("^[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[89ab][a-f0-9]{3}-?[a-f0-9]{12}")
	filename := path.Base(encryptedfilepath)
	uuid := strings.Split(filename, ".")[0]
	if !r.MatchString(uuid) {
		fmt.Printf("File name does not appear to be valid.\nPlease use the exact file name from the job status endpoint (i.e., of the format: <UUID>.ndjson).\n")
		os.Exit(1)
	}
}

func decryptCipher(ciphertext []byte, key *[32]byte) (plaintext []byte, err error) {
	block, err := aes.NewCipher(key[:])
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	if len(ciphertext) < gcm.NonceSize() {
		return nil, errors.New("malformed ciphertext")
	}
	return gcm.Open(nil,
		ciphertext[:gcm.NonceSize()],
		ciphertext[gcm.NonceSize():],
		nil,
	)
}

func decryptFile(privateKey *rsa.PrivateKey, encryptedKey []byte, filename string) {
	base := path.Base(filename)
	decryptedKey, err := rsa.DecryptOAEP(
		sha256.New(), rand.Reader, privateKey, encryptedKey, []byte(base))
	if err != nil {
		fmt.Println("Failed to decrypt encrypted key. Error:", err)
		os.Exit(1)
	}

	ciphertext, err := ioutil.ReadFile(filepath.Clean(filename))
	if err != nil {
		fmt.Println("Failed to read encrypted file. Error:", err)
		os.Exit(1)
	}

	var plaintext []byte
	key := [32]byte{}
	copy(key[:], decryptedKey[0:32])
	plaintext, err = decryptCipher(ciphertext, &key)
	if err != nil {
		fmt.Println("Failed to decrypt file. Error:", err)
		os.Exit(1)
	}

	fmt.Printf("%s", plaintext)
}

func getPrivateKey(loc string) *rsa.PrivateKey {

	pkFile, err := os.Open(filepath.Clean(loc))
	if err != nil {
		fmt.Println("Failed to open private key. Error:", err)
		os.Exit(1)
	}

	pemfileinfo, _ := pkFile.Stat()
	var size = pemfileinfo.Size()
	pembytes := make([]byte, size)
	buffer := bufio.NewReader(pkFile)

	_, err = buffer.Read(pembytes)
	if err != nil {
		fmt.Println("Failed to read private key. Error:", err)
		os.Exit(1)
	}

	data, _ := pem.Decode([]byte(pembytes))
	err = pkFile.Close()
	if err != nil {
		fmt.Println("Failed to close private key. Error:", err)
		os.Exit(1)
	}

	imported, err := x509.ParsePKCS1PrivateKey(data.Bytes)
	if err != nil {
		fmt.Println("Failed to parse private Key as PKCS1. Error:", err)
		os.Exit(1)
	}

	return imported
}

func main() {
	ek, err := hex.DecodeString(encryptedKey)
	if err != nil {
		fmt.Println("Failed to decode encrypted key. Error:", err)
		os.Exit(1)
	}
	pk := getPrivateKey(private)
	decryptFile(pk, ek, encryptedfilepath)
}

File: decryption_utils/Python/decrypt.py
from __future__ import print_function

import argparse
import binascii
import os
import re
import sys
import codecs

from argparse import RawTextHelpFormatter

try:
    from Crypto.Cipher import PKCS1_OAEP, AES
    from Crypto.PublicKey import RSA
    from Crypto.Hash import SHA256
except ImportError:
    print("Install requirements first: pip install -r requirements.txt", file=sys.stderr)
    raise SystemExit

# Match nonce and tag sizes used by BCDA
# Golang defaults: https://golang.org/src/crypto/cipher/gcm.go#L146
GCM_NONCE_SIZE = 12
GCM_TAG_SIZE = 16


def init():
    parser = argparse.ArgumentParser(
        description='Decrypt a BCDA payload.',
        formatter_class=RawTextHelpFormatter
    )
    parser.add_argument(
        '--key', dest='key', type=str,
        help="encrypted symmetric key used for file decryption (hex-encoded string)"
    )
    parser.add_argument(
        '--file', dest='file', type=str,
        help="location of encrypted file"
    )
    parser.add_argument(
        '--pk', dest='pk', type=str,
        help="location of private key to use for decryption of symmetric key"
    )

    args = parser.parse_args()

    if not args.key or not args.file or not args.pk:
        print("missing argument(s)", file=sys.stderr)
        raise SystemExit

    if not valid_uuid(args.file):
        print("""File name does not appear to be valid.
Please use the exact file name from the job status endpoint (i.e., of the format: <UUID>.ndjson).""",
              file=sys.stderr)
        raise SystemExit

    return args


def decrypt_cipher(ct, key):
    nonce = ct.read(GCM_NONCE_SIZE)
    cipher = AES.new(key, AES.MODE_GCM, nonce=nonce, mac_len=GCM_TAG_SIZE)
    ciphertext = ct.read()
    return cipher.decrypt_and_verify(
        ciphertext[:-GCM_TAG_SIZE],
        ciphertext[-GCM_TAG_SIZE:]
    )


def decrypt_file(private_key, encrypted_key, filepath):
    base = os.path.basename(filepath)
    cipher = PKCS1_OAEP.new(key=private_key, hashAlgo=SHA256, label=base.encode('utf-8'))
    decrypted_key = cipher.decrypt(encrypted_key)

    with open(filepath, 'rb') as fh:
        result = decrypt_cipher(fh, decrypted_key).decode("utf-8")

    print(result)


def get_private_key(loc):
    with open(loc, 'r') as fh:
        return RSA.importKey(fh.read())

def valid_uuid(filename):
    path, name = os.path.split(filename)
    uuid = name.split(".")[0]
    regex = re.compile('^[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[89ab][a-f0-9]{3}-?[a-f0-9]{12}\Z', re.I)
    match = regex.match(uuid)
    return bool(match)

def main():
    args = init()
    if sys.version_info[0] < 3:
        sys.stdout = codecs.getwriter('utf-8')(sys.stdout)
    ek = binascii.unhexlify(args.key)
    pk = get_private_key(args.pk)
    decrypt_file(pk, ek, args.file)

if __name__ == "__main__":
    main()